### PR TITLE
feat: improve handling of "abbreviated" flake refs

### LINKED ISSUE #20: `git`/`path` disambiguation
I wondered why `/path/to/flake` was being parsed as `git+file://` by nix.
After some digging i found that nix is "pre-parsing" the flakeref url [1] where it resolves the shortcuts `flake:`, `path:` and `git+file`, constructs a new url from those and only then parses the  new url.

Notably, since `path` and `git+file` urls are somewhat overlapping (both represent a local path) nix is actually trying to resolve the flakeroot and git dir to disambiguate.

The runix impl is substituting urls as part of the parsing but does _not_ perform any io, flake resolution.

In fact, i find it a bit out of order to do io for parsing flakerefs as that would make flakeref parsing dependent on the system state its performed on.

The question is how we should handle `/path/to/flake` urls if not resolving flakes.

My approach would be to add a `FlakeRefSource` that catches parses a `PathBuf` and query as opposed to a url. It would then have an impure method to resolve to either Path or Git Flakeref.
I do wonder though whether that should be a `FlakeRef` variant or just a separate struct.

As a variant, it would likely need to fail serialization due to ambiguity!

[1] https://github.com/flox/nix/blob/master/src/libexpr/flake/flakeref.cc#L125-L186

@mkenigs note:
A bit of an optimization, we may want to pull in for a Rust sprint

### PR DESCRIPTION
**Pull request by ysndr**
If not well defined, i.e. if the flakeref cannot be parsed as a url,
we resolve it either as an indirect flake or local path.

Note: if not "well-defined" parsing flakerefs is "impure",
      i.e. depends on the state of the local system (files).
      The resulting flakeref however, serializes into well-defined form.

Resolving an abbreviated URL to local files:

Nix supports referring to local files/paths without an explicit scheme.
However, it distinguishes between flakes in git repositories
(disambiguated using `git+file://`) from "free" ones (`path:`).
The logic for this in nix can be found at <https://github.com/NixOS/nix/blob/bf7dc3c7dc24f75fa623135750e8f10b8bcd94f9/src/libexpr/flake/flakeref.cc#L112C5-L197>
This method implements a subset of this logic assuming we require the path to exist and to be a flake.
In abstract we
1. locate a flake (by finding a `flake.nix` in all ancestors)
  - fail if directory does not exist
  - fail if skipping file system boundaries
  - fail if "exiting" a git repository (if there is one)
2. locate the root of the "containing" git repo (if applicable)
  - check whether `.git/shallow` exists and track in the url params
  - ensure there is no `?dir=` param if flake is in a subdir
3. construct a `file+git:` or `path:` url as required

Additionally this PR refactors the parsing of flakerefs in https://github.com/flox/runix/commit/5c91d2a4f3b2752d6ecac6127311cad2a3d3f4c8 to be URL first.

So far, all flakeref parsers implemented the parsing as `FromStr` only.
In all cases `FromStr` would first parse the input as `Url` and then process mainly the Url.
After this refactoring all flakerefs can only be _purely_ parsed using Urls or "well-defined" url strings (see above).
`FromStr` is now uniformly implemented as

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let url = Url::parse(s)?;
        Self::from_url(url)
    }

for all `FlakeRef` types.

If "impure" resolution is necessary, `FlakeRef::resolve_local` provides an interface to create a well-defined Url.
`FlakeRef::from_str` by always tries to impurely resolve urls for inputs that are not well-defined.

The change makes `FlakeRef` instances pure by default, which simplifies the parsing as impure components are stripped out.

### CODE REVIEW

File: crates/runix/src/flake_ref/mod.rs [hunk 1]
```diff
@@ -52,8 +67,33 @@ pub enum FlakeRef {
...
+            Err(_) => {
+                let resolved = if FLAKE_ID_REGEX.is_match(s) {
+                    Url::parse(&format!("flake:{s}")).map_err(|e| {
+                        ParseFlakeRefError::Indirect(indirect::ParseIndirectError::Url(e))
+                    })?
+                } else {
+                    FlakeRef::resolve_local(s)?
+                };
+
+                info!("Could not parse flakeref as URL, resolved locally to '{resolved:?}'");
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/mod.rs]:
```suggestion
                info!("Could not parse flakeref as URL; resolved locally to '{resolved:?}'");
```

File: crates/runix/src/flake_ref/mod.rs [hunk 2]
```diff
@@ -92,6 +132,150 @@ impl FromStr for FlakeRef {
...
+            if ancestor.join("flake.nix").exists() {
+                debug!("found for flake in {ancestor:?}");
+                break ancestor;
+            }
+
+            if path.join(".git").exists() {
+                Err(ResolveLocalRefError::GitRepoBoundary(
+                    ancestor.to_path_buf(),
+                ))?;
+            }
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/mod.rs:211]: Is path the same here every iteration?

Is this the same behavior as `nix`? Do you want to fail if there's a git submodule? Submodules have a `.git` file in them

File: crates/runix/src/flake_ref/mod.rs [hunk 3]
```diff
@@ -92,6 +132,150 @@ impl FromStr for FlakeRef {
...
+        };
+
+        // if there is git
+        if let Some(git_root) = flake_root.ancestors().find(|ancestor| {
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/mod.rs:215]: Optional: my intuition is that having a single loop would be more readable than looping through `ancestors()` twice

**Review by mkenigs — CHANGES_REQUESTED** [review]
I found this a little hard to review as it seems like there's a lot of changes in 5c91d2a4f3b2752d6ecac6127311cad2a3d3f4c8 but I'm not sure how that relates to this PR, could you briefly explain that (here or in the commit)? Or even split into a separate PR if that would be easy, but no worries if not

File: crates/runix/src/flake_ref/mod.rs (hunk 2 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/mod.rs:211]: ancestors iterates over all parent paths starting with the original path: https://doc.rust-lang.org/std/path/struct.Path.html#method.ancestors

As far as i can tell nix does not care too much about submodules. This part is the equivalent to https://github.com/NixOS/nix/blob/bf7dc3c7dc24f75fa623135750e8f10b8bcd94f9/src/libexpr/flake/flakeref.cc#L135-L136
That is if your submodule does not contain a flake nix would raise an error

File: crates/runix/src/flake_ref/mod.rs (hunk 3 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/mod.rs:215]: Again, this mirrors the nix implementation.

I thought it might actually be easier to read this way because you separate the flake "finding" from the git "finding". It's a bit confusion that we are also looking for the git boundary during the flake search but well..
Btw. the two loops are already "kind of" a single loop as the full set of ancestors is iterated until we find a flake and then possibly completed when we search the git root.

We can refactor this a bit if you'd like but i think its better to stick close to the reference for now..

Comment (tomberek) [reply crates/runix/src/flake_ref/mod.rs:215]: Looks pretty close to the upstream logic.

Comment (mkenigs) [reply crates/runix/src/flake_ref/mod.rs:215]: Nope I don't mind

File: crates/runix/src/flake_ref/mod.rs (hunk 2 above)
Comment (mkenigs) [reply crates/runix/src/flake_ref/mod.rs:211]: Right but in line 201 you aren't using `ancestor` you're using `path` - how is that changing?

Comment (ysndr) [reply crates/runix/src/flake_ref/mod.rs:211]: Ah I see, that was wrong it should indeed be `ancestor`, thanks

File: crates/runix/src/flake_ref/file.rs [hunk 4]
```diff
@@ -187,61 +189,25 @@ impl<Protocol: FileProtocol, Type: ApplicationProtocol> FlakeRefSource
...
-            "{scheme_with_application}:",
-            scheme_with_application = Self::scheme()
-        )) {
+    fn parses(maybe_ref: &Url) -> bool {
+        // let url = if let Ok(url) = Url::parse(maybe_ref) {
+        //     url
+        // } else {
+        //     return false;
+        // };
+
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/file.rs]: Do you mean to leave this in?

File: crates/runix/src/flake_ref/git.rs [hunk 5]
```diff
@@ -169,4 +195,53 @@ mod tests {
             flakeref
         );
     }
+
+    /// assert that relative file urls are
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/git.rs]: Are...?

File: crates/runix/src/flake_ref/git.rs [hunk 6]
```diff
@@ -169,4 +195,53 @@ mod tests {
...
+            url.url.path(),
+            std::env::current_dir()
+                .unwrap()
+                .parent()
+                .unwrap()
+                .to_string_lossy()
+        );
+        dbg!(url.to_string());
+
+        // with "//" an absolute path is required prt the Url spec for file urls
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/git.rs]: prt?

File: crates/runix/src/flake_ref/git.rs [hunk 7]
```diff
@@ -169,4 +195,53 @@ mod tests {
...
+        let url = GitRef::<protocol::File>::from_str("git+file:///var/www").unwrap();
+        dbg!(&url);
+        assert_eq!(url.url.path(), "/var/www");
+        dbg!(url.to_string());
+
+        // with "//" an absolute path is required prt the Url spec for file urls
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/git.rs]: prt?

File: crates/runix/src/flake_ref/git.rs [hunk 8]
```diff
@@ -169,4 +195,53 @@ mod tests {
...
+                .parent()
+                .unwrap()
+                .to_string_lossy()
+        );
+        dbg!(url.to_string());
+
+        // with "//" an absolute path is required prt the Url spec for file urls
+        let url = GitRef::<protocol::File>::from_str("git+file:///var/www").unwrap();
+        dbg!(&url);
+        assert_eq!(url.url.path(), "/var/www");
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/git.rs:229]: (can ignore) any reason not to follow a pattern more similar to that in `to_from_json` where you assert everything is equal (rather than just `url.url.path`?)

File: crates/runix/src/flake_ref/mod.rs [hunk 9]
```diff
@@ -52,46 +67,221 @@ pub enum FlakeRef {
...
+
+                if device != original_device {
+                    Err(ResolveLocalRefError::FileSystemBoundary(
+                        ancestor.to_path_buf(),
+                    ))?
+                }
+            }
+
+            if ancestor.join("flake.nix").exists() {
+                debug!("found for flake in {ancestor:?}");
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/mod.rs]:
```suggestion
                debug!("found flake in {ancestor:?}");
```

File: crates/runix/src/flake_ref/mod.rs [hunk 10]
```diff
@@ -286,5 +490,126 @@ pub(super) mod tests {
...
+    #[test]
+    fn test_resolve_absolute_git_local() {
+        let flake_test_dir = tempfile::tempdir().unwrap();
+
+        let git_dir = flake_test_dir
+            .path()
+            .canonicalize()
+            .unwrap()
+            .join("withgit");
+        fs::create_dir_all(&git_dir).unwrap();
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/mod.rs:549]:
```suggestion
```

File: crates/runix/src/flake_ref/mod.rs [hunk 11]
```diff
@@ -286,5 +490,126 @@ pub(super) mod tests {
...
+                .unwrap()
+                .to_string(),
+            Url::parse(&format!("git+file:{git_dir_string}?dir={flake_dir_name}"))
+                .unwrap()
+                .to_string()
+        )
+    }
+
+    #[test]
+    fn test_resolve_relative_git_local() {
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/mod.rs:601]: Combine this with `test_resolve_absolute_git_local`? The setup is the same and then you can assert that all three things are equal

File: crates/runix/src/flake_ref/mod.rs [hunk 12]
```diff
@@ -286,5 +490,126 @@ pub(super) mod tests {
             dbg!(FlakeRef::from_str("flake:nixpkgs")).unwrap(),
             FlakeRef::Indirect(_)
         ));
+        assert!(matches!(
+            dbg!(FlakeRef::from_str("nixpkgs")).unwrap(),
+            FlakeRef::Indirect(_)
+        ));
+    }
+
+    #[test]
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/mod.rs:501]: Optional
```suggestion
    /// basic
    /// └── flake.nix
    #[test]
```

File: crates/runix/src/flake_ref/mod.rs [hunk 13]
```diff
@@ -286,5 +490,126 @@ pub(super) mod tests {
...
+            FlakeRef::resolve_local(&path.to_string_lossy())
+                .unwrap()
+                .to_string(),
+            Url::parse(&format!("path:{}", basic.to_string_lossy()))
+                .unwrap()
+                .to_string()
+        )
+    }
+
+    #[test]
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/mod.rs:540]:
```suggestion
    /// withgit
    /// ├── .git
    /// └── flake.nix
    #[test]
```

File: crates/runix/src/flake_ref/mod.rs [hunk 14]
```diff
@@ -286,5 +490,126 @@ pub(super) mod tests {
...
+            FlakeRef::resolve_local(&git_dir_string)
+                .unwrap()
+                .to_string(),
+            Url::parse(&format!("git+file:{git_dir_string}"))
+                .unwrap()
+                .to_string()
+        )
+    }
+
+    #[test]
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/mod.rs:569]:
```suggestion
    /// withgit
    /// ├── .git
    /// └── inner
    ///     └── flake.nix
    #[test]
```

File: crates/runix/src/flake_ref/mod.rs [hunk 15]
```diff
@@ -286,5 +490,126 @@ pub(super) mod tests {
...
+        let relative_path = pathdiff::diff_paths(&git_dir, env::current_dir().unwrap()).unwrap();
+
+        assert_eq!(
+            FlakeRef::resolve_local(relative_path.to_string_lossy())
+                .unwrap()
+                .to_string(),
+            Url::parse(&format!("git+file:{}", git_dir.to_string_lossy()))
+                .unwrap()
+                .to_string()
+        )
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/mod.rs:622]: There isn't coverage for any of the error edge cases - crossing git boundary, crossing device boundary, inconsistent dir param

**Review by mkenigs — APPROVED** [review]
Looks good - thanks for adding tests! Left a bunch of nits, the only thing I'd really suggest considering is the test coverage for errors

File: crates/runix/src/flake_ref/mod.rs (hunk 10 above)
Comment (mkenigs) [reply crates/runix/src/flake_ref/mod.rs:549]: Same thing elsewhere

File: crates/runix/src/flake_ref/file.rs (hunk 4 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/file.rs]: no "parses" requires a url now

File: crates/runix/src/flake_ref/mod.rs (hunk 11 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/mod.rs:599]: Not quite sure how to combine them without making a bit of an awkward setup: 

```
withgit
├── .git
├── flake.nix
└── inner
  └── flake.nix
```

I think i prefer testing those separately for now

File: crates/runix/src/flake_ref/git.rs (hunk 8 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/git.rs:229]: The test here is to demonstrate the particularities of `file://` urls

File: crates/runix/src/flake_ref/mod.rs (hunk 15 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/mod.rs:622]: Don't know how to account for device boundaries but i added tests for git boundary and dir param

File: crates/runix/src/flake_ref/mod.rs (hunk 11 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/mod.rs:601]: Not quite sure how to combine them without making a bit of an awkward setup: 

```
withgit
├── .git
├── flake.nix
└── inner
  └── flake.nix
```

I think i prefer testing those separately for now

File: crates/runix/src/flake_ref/git.rs [hunk 16]
```diff
@@ -169,4 +195,53 @@ mod tests {
             flakeref
         );
     }
+
+    /// assert that relative file urls are rosolved to git urls correctly
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/git.rs]:
```suggestion
    /// assert that relative file urls are resolved to git urls correctly
```

File: crates/runix/src/flake_ref/mod.rs [hunk 17]
```diff
@@ -286,5 +490,205 @@ pub(super) mod tests {
...
+            ResolveLocalRefError::InconsistentDirParam(_, _)
+        ));
+
+        // /<git_root>?dir=inner
+        // if not given in a well-defined format, paths must directly point to the flake
+        // or a folder inside the flake.
+        // nix does not prescribe using the `dir` parameter as an alternative root to search for flakes.
+        // So even though we parse
+        //    /<git_root>/inner -> git+file://<git_root>?dir=inner
+        // It is not possible to reger to an inner path using a `dir` parameter in a non-well-defined url.
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/mod.rs]: `reger`?

File: crates/runix/src/flake_ref/mod.rs [hunk 18]
```diff
@@ -286,5 +490,205 @@ pub(super) mod tests {
...
+        assert!(matches!(
+            dbg!(
+                FlakeRef::resolve_local(format!("{}?dir=other", flake_root.to_string_lossy()))
+                    .expect_err("should fail")
+            ),
+            ResolveLocalRefError::InconsistentDirParam(_, _)
+        ));
+
+        // /<git_root>/inner?dir=inner
+        // nix does resolves the path to
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/mod.rs]:
```suggestion
        // nix resolves the path to
```

File: crates/runix/src/flake_ref/mod.rs (hunk 17 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/mod.rs]:
```suggestion
        // It is not possible to refer to an inner path using a `dir` parameter in a non-well-defined url.
```
